### ISSUE TITLE
Add setting to require existing password when changing password

### ISSUE THREAD

**Issue by Sytten** [ory/kratos#2586]
### Preflight checklist

- [X] I could not find a solution in the existing issues, docs, nor discussions.
- [X] I agree to follow this project's [Code of Conduct](https://github.com/ory/kratos/blob/master/CODE_OF_CONDUCT.md).
- [X] I have read and am following this repository's [Contribution Guidelines](https://github.com/ory/kratos/blob/master/CONTRIBUTING.md).
- [ ] This issue affects my [Ory Cloud](https://www.ory.sh/) project.
- [ ] I have joined the [Ory Community Slack](https://slack.ory.sh).
- [ ] I am signed up to the [Ory Security Patch Newsletter](https://ory.us10.list-manage.com/subscribe?u=ffb1a878e4ec6c0ed312a3480&id=f605a41b53).

### Describe your problem

This is a follow-up to https://github.com/ory/kratos/discussions/1971 since I don't think the solution is enough for certain uses cases (including regulated industries).
The privileged session is a good idea, but since it's not a manual operation from the user (aka they get a privileged session on login), this means that a client side vulnerability can still trigger a password change and an account take over.

### Describe your ideal solution

Adding a setting to require the current password for this password change flow mitigates the issue entirely.

### Workarounds or alternatives

None at the moment.

### Version

0.10.1

**Comment by alantbarlow** [comment]
I agree with this, there should be a way to require password to update any of the user details. If this feature is enabled, you can have a self-service page that lets the customer update their info but only be able to save the changes if the password is correct. It could even be implemented in one of a couple ways...

 1. asking for current password after "saving" the changes (but before persisting the changes) as a way of verification.
 2. adding a current password field that activates the "save changes" button, then it could check the password for a match before persisting the changes.

This feature should be completely optional as it would not really be needed if going the passwordless route.

**Comment by vinckr** [comment]
You can change the [Authenticator Assurance Level](https://www.ory.sh/docs/guides/mfa/overview#authenticator-assurance-level-aal) required for settings/profile updates.

So if you set the aal level to strict, users dont get a privileged session from login automatically - they have to use a second factor. 

>client side vulnerability can still trigger a password change and an account take over.

This is prevented when you use a second factor - as are many other attack vectors.

If security is the motivation for this feature, I don't really see how entering the password (again) would be an improvement over using a second factor - but open to have my mind changed :)

**Comment by alantbarlow** [comment]
it sounds like the ability to change the aal level to prevent the automatic privileged session could solve the issue. The scenario of someone maliciously hacking your account trough the use of a connected social login method, can also be solved by including a second factor for authentication.

But this brings up two questions.

1. If you don't want to require a second factor for login, is it possible to only require the second factor when the user does actions that need privileged sessions (like when modifying account info or adding/removing login methods)?

2. if the user has multiple login methods assigned to their account, is it possible to require them to use a different method than the one they used to sign-in to grant a privileged session?

here is an example scenario... A company requires email/password sign-up and doesn't want to bother the user to use a second factor to login. Instead, the user is prompted to connect their social providers as additional options to login and as a security precaution to prevent account takeover.

In this scenario, if the user setup an additional login method and someone hacked their account by way of stealing their password. is it possible for the system to know what login method was used for the session and require an alternative method before granting a privileged session?

**Comment by Sytten** [comment]
Not stale

**Comment by Sytten** [comment]
@vinckr While I agree that in theory the `aal` is code without https://github.com/ory/kratos/issues/3668 I don't see how we would use it. Because we don't want to mandate all users to setup a 2FA.

**Comment by Sytten** [comment]
Not stale

**Comment by vinckr** [comment]
@Sytten 
As https://github.com/ory/kratos/issues/3668 and https://github.com/ory/kratos/issues/3494 are implemented, can you use that instead?

**Comment by Sytten** [comment]
I still dont think its a good idea to allow someone to change a password without knowing the existing password, literally everybody does that and for a good reason.

**Comment by Sytten** [comment]
Not stale

**Comment by vinckr** [comment]
Hello @Sytten 

can you briefly give more context on how the current implemented [privileged session](https://www.ory.sh/kratos/docs/concepts/session/#privileged-sessions) and [step-up auth](https://www.ory.com/docs/kratos/mfa/step-up-authentication) options don't satisfy this use case? 

>The privileged session is a good idea, but since it's not a manual operation from the user (aka they get a privileged session automatically on login), this means that a client side vulnerability can still trigger a password change and an account take over.

Maybe I'm not getting something but do log-in you need to know the password? And you can set the privileged session to something really low, like a couple of minutes. 
Also if that is not enough you can still use step-up auth and ALWAYS require the user to re-authenticate when they want to access the settings flow.

>since you get a privilege session from a social login (most of which won't ask for a password on login if the user is already logged-in with the provider), it is possible to change the password of a user without ever knowing any password of the targeted user.

Wouldn't that require the attacker to already have access to the social login account of the user? And if that is an attack vector maybe you should reconsider using social login at all? 
Or at the very least also require a second authN method on every login, for example through a physical key or TOTP, like @alantbarlow pointed out. 

I don't want to be contrarian just trying to suss out what the actual use case is behind this request and what the gap is/how it can be solved better. It seems to me there are already a couple of ways how this can be solved. 

Also to answer @alantbarlow questions: 

>If you don't want to require a second factor for login, is it possible to only require the second factor when the user does actions that need privileged sessions (like when modifying account info or adding/removing login methods)?

Yes that is possible and I think also the default setting if you enable MFA.

>if the user has multiple login methods assigned to their account, is it possible to require them to use a different method than the one they used to sign-in to grant a privileged session?
>here is an example scenario... A company requires email/password sign-up and doesn't want to bother the user to use a second factor to login. Instead, the user is prompted to connect their social providers as additional options to login and as a security precaution to prevent account takeover.

It probably is possible hacking around a bit, but on the other hand if it is such a "high-security" scenario why is it important to not "bother" the user with MFA? 
Why not stop using passwords altogether and use only social login or passkeys in this scenario and require MFA for more "sensitive" actions on a per case basis using [step-up auth](https://www.ory.com/docs/kratos/mfa/step-up-authentication)?

>In this scenario, if the user setup an additional login method and someone hacked their account by way of stealing their password. is it possible for the system to know what login method was used for the session and require an alternative method before granting a privileged session?

this is probably possible through events or logs, but requires some custom coding. 
But I think it isn't a realistic scenario - if their account is already pwned there is a good chance that their social login is also pwned. 
I think passkeys or MFA required for sensitive actions solves this better.

**Comment by Sytten** [comment]
> Wouldn't that require the attacker to already have access to the social login account of the user? And if that is an attack vector maybe you should reconsider using social login at all?

Not really. This makes an XSS vuln into account takeover vuln, since you can force the user to get into a re-up flow via social and then set a password in the background. Also for cases where you want to disable password auth "sometimes" (SSO only mode) this might help bypass it but I would need to double check.

IDK maybe its 20y of being on the internet but not asking for current password when setting a new password is sus outside of an explicit password reset flow.

And even just for compliance box checking you might need that.

**Comment by vinckr** [comment]
Thanks for the quick response @Sytten !

What about the alternative where you use a custom settings UI and only redirect the user with `?refresh=true` (or `?refresh=true&aal=aal2`)
You can add an additional check and only allow the user to see the settings UI when the session is only X seconds old. 

Btw is this connected to an actual use case that you have at the moment and you are prevented from using Ory Kratos bc of a missing compliance checkbox or is it more theoretical? 
I'm just asking because if it is we can probably help but also because across hundreds of use cases I worked on (also in highly regulated industries) this was never a hard compliance requirement - and if it was more important the custom UI route was always accepted as an adequate solution.

**Comment by Sytten** [comment]
I mean UI patch is not security if this is what you are proposing? This would get found in a proper pentest anyway. If this is not in your threat model that is fine and feel free to close the issue but this is very much an issue in my book.

We already use kratos, this issue is 3y old. Still thinks it's a bad design and there is a reason this is not the industry standard. GitHub has a system similar to aal2 and still asks for current password to change. I dont think its a crazy expectation.

**Comment by vinckr** [comment]
>UI patch is not security

I think you should be able to do the same on the API level (check if the session is older than X seconds) - but I guess the privileged sessions feature does the same...

To me the gist is that re-authentication is important for sensitive actions, not necessarily asking for the old password.

Google for example does the same - it asks me to authenticate again when I want to change the password - but I don't need to necessarily supply my old password, I can use any authentication method.

>feel free to close the issue

To be clear I also don't think it's a **crazy** feature request.  
It seems to me that it's a nice to have - evidently it didn't block you from using Ory Kratos in production (and no one else is looking for the feature as far as I can tell).

I'm mainly looking for steelman arguments why this should be prioritized over other features. I think that it will just stay open (arguably stale), because it doesn't improve security substantially.
Ultimately it's not my call if it gets implemented or not.

The only real use case that I can come up with is that it offers some protection against "Walk-up / shared device" attacks. But I think re-authentication also prevents those? 

I can't see anything about requiring old passwords in [NIST](https://pages.nist.gov/800-63-3/sp800-63b.html), but [OWASP](https://github.com/OWASP/ASVS/blob/master/5.0/en/0x15-V6-Authentication.md) does mention it (but the section in NIST they link to doesn't mention it 🤔)
What NIST recommend its covered by requiring MFA to access /settings and a mechanism like privileged sessions.

**Comment by Sytten** [comment]
I agree with the privileged session is a good design, but the Google example is not good since you cant social login into Google to then change the password of Google. That would be considered a crazy bad bypass I am pretty sure.

In the walk up example the privileged session doesn't help cause Google is already authenticated so the SSO just goes through and then you can directly change or set (if it wasnt used previously) the password.

For the sake of example let's say you steal a Discord cookie via a malicious bot. If your app allows discord SSO you can now login as privileged session and takeover that app account permanently too without user knowing (AFAIK kratos doesn't send an email on password change or set).

A bit convoluted I will admit but kinda shows that privileged session doesn't work all that well with social providers that are already logged in and approved.
